Core Sketcher: Bounding box rework

PaddleStroke opened the pull request:
This PR does several things.

A - Port changes from realthunder's branch.
Principal : https://github.com/realthunder/FreeCAD/commit/cb519540688e35d3fdf2087dd14fbe42fd37d5cb
A a few additional fixes : 
https://github.com/realthunder/FreeCAD/commit/d55138b0d27a9aecfa9111cf0d520f1d0e3cd6b5
https://github.com/realthunder/FreeCAD/commit/8a0484f2493f3cd14ced4723b91c99c87bbb4db8
https://github.com/realthunder/FreeCAD/commit/a82fc4effe2794cf1f136e2d211558302adfff1b
https://github.com/realthunder/FreeCAD/commit/6a454ed5a13ffcd93403b68cf30affd2bfec640e
https://github.com/realthunder/FreeCAD/commit/68f794119867077c3cf7b2e2187b81c040f04053

Note: not all of the content of those commits where ported. Only the necessary parts.

This port does : 
- fix https://github.com/FreeCAD/FreeCAD/issues/20584
- fix https://github.com/FreeCAD/FreeCAD/issues/16467
- fix https://github.com/FreeCAD/FreeCAD/issues/25172

Additionally this port is giving access to a viewObjects() function which enables us to 
- fix https://github.com/FreeCAD/FreeCAD/issues/25950

Last I also added a 
- fix https://github.com/FreeCAD/FreeCAD/issues/14007

Summary of the fixes : 
- BoundingBox property : the bbox now has the correct global placement
- Fit selection: Works with elements (faces, edges...) for both features and links.
- Fit selection: now works for sketches while in edit. And works with individual elements (faces, edges).
- Entering sketch edit mode : now the sketch is fit to the selection. Which avoids opening on empty space at random positions

kadet1090 commented on src/Gui/View3DInventorViewer.cpp:
```diff
@@ -3786,6 +3922,52 @@ void View3DInventorViewer::alignToSelection()
+    cam->viewAll(&path, getSoRenderManager()->getViewportRegion());
+    path.truncate(0);
+    pcGroup->unref();
+#endif
```
@FreeCAD/maintainers do we need to support coin 3?

PaddleStroke replied:
Let me know if I should remove it

kadet1090 commented on src/Gui/ViewProvider.cpp:
```diff
@@ -1112,70 +1117,231 @@ void ViewProvider::setRenderCacheMode(int mode)
+        if (m) {
+            mat = *m;
+        }
+    }
```
No 1 letter variable names please, we have space to store meaningful names.

kadet1090 commented on src/Gui/ViewProviderGeometryObject.cpp:
```diff
@@ -95,23 +95,24 @@ ViewProviderGeometryObject::ViewProviderGeometryObject()
-    pcBoundingBox->unref();
-    pcBoundColor->unref();
-    pickStyle->unref();
+    if (pcShapeMaterial) {
```
If you use `CoinPtr` these unrefs won't be needed

kadet1090 commented on src/Gui/3Dconnexion/navlib/NavlibPivot.cpp:
```diff
@@ -1,295 +1,295 @@
-/***************************************************************************
```
I don't think this file should be changed?

kadet1090 reviewed (commented):
Few comments, mostly about dead code.

PaddleStroke commented on src/Gui/ViewProviderGeometryObject.cpp:
```diff
@@ -300,26 +344,45 @@ void ViewProviderGeometryObject::showBoundingBox(bool show)
+        pcSwitchSensor->setData(this);
+        pcSwitchSensor->attach(pcModeSwitch);
+        pcSwitchSensor->setFunction([](void* data, SoSensor*) {
+            reinterpret_cast<ViewProviderGeometryObject*>(data)->addBoundSwitch();
```
I do not know, do you?

kadet1090 replied:
In general for conversion from/to `void*` the `static_cast` should be used - especially if we know the type. reinterpret_cast should be used if casting between different types (let's say `struct { int number; }` and `int` which can be _interpreted_ - hence reinterpret_cast - from the same memory as different data types).

hyarion commented on src/App/ComplexGeoData.cpp:
```diff
@@ -734,4 +734,11 @@ const std::string ComplexGeoData::dumpElementMap() const
+        return name + elementMapPrefix().size();
+    }
+    return nullptr;
+}
```
would it be an option to use `std::string_view` instead of pointer based c-strings?
```suggestion
std::string_view ComplexGeoData::isMappedElement(std::string_view name)
{
    if (std::string_view prefix = elementMapPrefix(); name.starts_with(prefix)) {
        return name.substr(prefix.size());
    }
    return {};
}
```

hyarion commented on src/Base/Tools.h:
```diff
@@ -173,6 +173,14 @@ inline T fmod(T numerator, T denominator)
+    std::hash<T> hasher;
+    seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+}
+
```
I don't think you are allowed to copy code like this without including the license for it.

if we can't use boost directly we might get away with including the license notice like this:
```diff
   // copied from boost::hash_combine.
+++// Copyright 2005-2014 Daniel James.
+++// Copyright 2021, 2022, 2025 Peter Dimov.
+++// Distributed under the Boost Software License, Version 1.0.
+++// https://www.boost.org/LICENSE_1_0.txt
   template<class S, class T>
```

PaddleStroke replied:
Oh actually this is in Part module. So I cannot use it in Gui.
So I suggest just moving it from TopoShapeMapper to Tools. If I only move it do I still need to tackle this ?

PaddleStroke commented:
bump

PaddleStroke commented:
The review comments have been addressed

PaddleStroke commented:
Conflict from https://github.com/FreeCAD/FreeCAD/pull/28982 has been fixed.
Please do not let this rot again...

tritao commented:
I think the overall direction here is good. IMHO FreeCAD does need a higher-level bounding-box concept in the view-provider layer, rather than relying entirely on Coin3D.

That said, I don’t think this PR is in a mergeable state yet.

This is a large behavior change across bbox calculation, viewer fit behavior, selection rendering, links, CAM, and Sketcher UX, but I don’t see any corresponding test coverage added for those paths. Given the size of the change, I would expect at least targeted tests for placement/link/subobject bbox correctness, bbox cache invalidation, `ViewSelection`/`viewObjects()`, and the Sketcher edit-mode fitting behavior. Without that, it is hard to validate this safely.

I hope mantainers can start to be stricter when it comes to merging PRs without automated tests, especially major pieces of core functionality like this.

There also appear to be concrete correctness problems in the current implementation:

 - The new bounding-box cache looks unsafe to me. The result now depends on the active viewer / viewport, but the cache key does not. That means a bbox computed in one 3D view could be reused later in a different view or after a resize, even though the inputs are no longer really the same.

 - The reusable bounding-box action also seems to keep some state between calls that is not fully cleaned up. In the `transform == false` path, a reset path is installed, but I do not see it being cleared afterward. That makes me worry that one bbox query can accidentally affect the next one on the same thread.

 - The CAM-specific `ViewProviderPath::_getBoundingBox()` change looks broken at the type/signature level. The function signature no longer matches the base virtual, so it seems this method is no longer actually overriding anything. If that is correct, then the new CAM bbox code is never called.

 - `viewAll(float factor)` also looks off to me. The new code seems to grow the box by the full width/height/depth around the center, which would make the effective scaling larger than the requested factor.

If you'd like some help with tests for this feel free to ping me on Discord.

PaddleStroke commented:
> The new bounding-box cache looks unsafe to me. The result now depends on the active viewer / viewport, but the cache key does not. That means a bbox computed in one 3D view could be reused later in a different view or after a resize, even though the inputs are no longer really the same.

As mentioned before, this caching mechanism can be removed. It was bundle in the fixes made by RealThunder, but I don't see much value in them as I don't see what issue this is fixing. I think it's probably if you use some kind of 'show bounding box on selection' that I seem to recall exist in linkstage (it may even exist in main).

@tritao please check if your claims about the cache mechanism are correct. Because they appear to be guess work more than actual issues. Then confirm if this cache mechanism should be dropped or not. Again I'm perfectly fine with dropping it, but just that we don't drop it for nothing.

For instance : 

> That means a bbox computed in one 3D view could be reused later in a different view

Why is this a problem ? An object shape is the same in all the 3D views of it. So the bbox is the same...

Appears to be working : 


https://github.com/user-attachments/assets/e562a89d-c7a8-4b05-b318-8c8ede5cfa42

Changing the size also works correctly.

PaddleStroke commented:
So I have finished to look at those 4 points @tritao. Please check below : 

> The new bounding-box cache looks unsafe to me. The result now depends on the active viewer / viewport, but the cache key does not. That means a bbox computed in one 3D view could be reused later in a different view or after a resize, even though the inputs are no longer really the same.

For this I'm not sure. I cannot find any actual issue. So please let me know if I should remove the caching or not. I'm ok both ways.

> The reusable bounding-box action also seems to keep some state between calls that is not fully cleaned up. In the transform == false path, a reset path is installed, but I do not see it being cleared afterward. That makes me worry that one bbox query can accidentally affect the next one on the same thread.

This is correct. I have pushed a fix for this.

> The CAM-specific ViewProviderPath::_getBoundingBox() change looks broken at the type/signature level. The function signature no longer matches the base virtual, so it seems this method is no longer actually overriding anything. If that is correct, then the new CAM bbox code is never called.

This is correct. It was a typo while porting. I fixed the type.

> viewAll(float factor) also looks off to me. The new code seems to grow the box by the full width/height/depth around the center, which would make the effective scaling larger than the requested factor.

This is also correct. I had not identified this issue because `viewAll(float factor)` function is never used in the source with factor != 1. Anyway if you did call the function from python like this : `Gui.ActiveDocument.ActiveView.fitAll(0.5)` it was indeed wrong. I have corrected the math error.

hyarion commented:
@PaddleStroke I agree with you, dropping the caching here might be the right call. It would simplify the code and we don't know why it was added in the first place so it could be a premature optimization that will just bite us in the future.

PaddleStroke commented:
What about deactivating it by default? There's this UseBoundingBoxCache parameter. This way if we need the optimization later on we'll have it. Or if someone wants to use it.

tritao commented:
> @tritao please check if your claims about the cache mechanism are correct. Because they appear to be guess work more than actual issues. Then confirm if this cache mechanism should be dropped or not. Again I'm perfectly fine with dropping it, but just that we don't drop it for nothing.
>
> Why is this a problem ? An object shape is the same in all the 3D views of it. So the bbox is the same...

I think we should distinguish between two different concepts here:

- geometry bbox
- render / scene-graph bbox

You are correct that a geometry bbox should be the same in all 3D views. But a render / scene-graph bbox can be view-dependent, for example when nodes react to zoom level or viewport size (`SoAutoZoomTranslation`, `SoZoomTranslation`).

My understanding is that this PR is aiming for a geometry-based bbox at the FreeCAD level, which I agree with. However, in the current implementation many of the leaf computations still go through `SoGetBoundingBoxAction`, and that path is seeded from the current viewer / viewport.

That is why I am still uneasy about the cache in its current form. It assumes the bbox only depends on the object, subname, and transform, while the implementation still seems able to traverse scene-graph nodes whose bbox may depend on the active view.

I cannot say at this point whether this turns into an observable bug in practice. But since those edge cases do not seem to have been clearly thought through, I would still prefer the safer option of dropping the cache for now.

hyarion commented:
> What about deactivating it by default? There's this UseBoundingBoxCache parameter. This way if we need the optimization later on we'll have it. Or if someone wants to use it.

If you want you could split the PR so you have the bounding box rework as one PR that is easier to merge, then another PR with the caching which could be a draft if anyone wants to check it out.

PaddleStroke commented:
I will not attempt to get this mechanism in later or in another PR. Either I leave it in and disable it or something, or I drop it entirely and won't come back to it.

So I just drop it?

PaddleStroke commented:
So if no one object by tomorrow I will just drop the cache from the PR so that this can be merged.

PaddleStroke commented:
@hyarion the bbox caching has been removed.
Conflicts solved

PaddleStroke commented:
Conflict solved. Please check this out @sliptonic @hyarion

PaddleStroke commented:
bump

PaddleStroke commented:
Anything holding this one ?

PaddleStroke commented:
Bump

tritao commented on src/Base/Tools.h:
```diff
@@ -174,6 +174,18 @@ inline T fmod(T numerator, T denominator)
+// Distributed under the Boost Software License, Version 1.0.
+// https://www.boost.org/LICENSE_1_0.txt
+template<class S, class T>
+inline void hash_combine(S& seed, const T& v)
```
```suggestion
inline void hashCombine(S& seed, const T& v)
```

PaddleStroke replied:
I'm only moving this helper from Part to Base.
Which was needed because of the caching was using this helper.
I have not reverted the move of this because I think it makes more sense to have it in Base/Tools than in Part.
But I can revert this change.
I don't see the point of renaming it, especially since I'm just moving it.

PaddleStroke replied:
Let me know if this is blocking for you. If so I will revert those changes. Or I guess I can rename the function as well. Whatever you prefer.

tritao replied:
Didn't realize it was just a move, in that case it seems acceptable to leave the old name.

tritao commented on src/Gui/Tree.cpp:
```diff
@@ -5910,7 +5910,11 @@ void DocumentItem::selectItems(SelectionReason reason)
         }
         else {
-            getTree()->syncView(newSelect->object());
+            // Here the selection change is most likely trigger from some thing
```
What's the point here of adding commented code? If it's not needed anymore then it should just be removed I think.

PaddleStroke replied:
I seem to recall that it was cherry picked from RT, he probably had his reason to leave it behind. I removed it.

hyarion commented:
There are some areas where we could make this a bit neater code wise but while discussing it with OP we concluded that it would require some other refactoring making it out of scope for this PR

PaddleStroke commented:
For reference the refactor is this : https://github.com/FreeCAD/FreeCAD/pull/29991#issuecomment-4613717913

PaddleStroke commented:
CI was failing because bool force was unused after trimming the bbox caching. So I removed that parameter since it's not used anymore

sliptonic reviewed (approved):
affects nothing in tool path calculation.  Might subtly affect view fitting with path arrows.

PaddleStroke commented:
CI passed @hyarion

PaddleStroke commented:
Anything holding this up?

tritao reviewed (commented):
Looks mostly good, I think some parts of the code still can be improved, but probably easier to send a follow-up later PR for this, just some minor comments that I had pending here from 2 weeks ago I forgot to submit.

PaddleStroke commented:
Ok so now are we all good here?

kadet1090 commented:
@chennes is going to test it with the test suite

PaddleStroke commented:
Two more conflicts this week.

https://github.com/FreeCAD/FreeCAD/commit/63f4c2f4fe14ab235e1129d2621ca3b4735716f3
https://github.com/FreeCAD/FreeCAD/commit/7638bf38aacec723c1d09318158dc69e60177b3e

The one from werner is trickier, the code is gone. Guess I'll have to dive in.

PaddleStroke commented:
Conflict solved

chennes commented:
> @chennes is going to test it with the test suite

No regressions with the test suite, everything comes back clean: old files recompute perfectly with these changes.

PaddleStroke commented:
Great thanks.
Anything else holding this back ?

hyarion commented:
@PaddleStroke I cannot merge this due to a conflict. Can you resolve it and DM me on discord and I'll merge it.

PaddleStroke commented:
<img width="2007" height="299" alt="image" src="https://github.com/user-attachments/assets/877ab57e-0de7-469b-b938-a7e606b5ccee" />
Hmm not seeing the conflict

PaddleStroke commented:
I clicked update button.

PaddleStroke commented:
Ok after the 8 months the PR history is such a mess. Why github has this `Update branch` button I wonder. Especially since it's the default button.

realthunder committed:
Gui: improve view provider bounding box calculation

PaddleStroke committed:
Sketcher: enable fit selection while in edit

PaddleStroke committed:
Sketcher: Fit sketch content when starting edit.

PaddleStroke commented:
History fixed. Rebased on latest main

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
